File: 2021/1.hs
part1 :: [Int] -> Int
part1 xs = length . filter (uncurry (<)) . zip xs $ (drop 1 xs)

part2 :: [Int] -> Int
part2 xs = part1 $ map (\(a,b,c) -> a+b+c) $ zip3 xs (drop 1 xs) (drop 2 xs)

main = do input <- getContents
          let xs = map read . lines $ input in
            (putStrLn . part1 $ xs) >> (putStrLn . part2 $ xs)

File: 2024/1.rb
input = ARGF.read.lines
ax = []
bx = []

input.each do |line|
    a, b = line.split.map(&:to_i)
    ax << a
    bx << b
end

ax.sort!
bx.sort!

part1 = ax.zip(bx).map do |(a, b)|
    (b - a).abs
end.sum

puts part1

bf = bx.group_by(&:itself).transform_values(&:size)
part2 = ax.map do |a|
    a * bf.fetch(a, 0)
end.sum

puts part2